scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -42,37 +55,102 @@ class KRouter[A](val config: Config[A], val network: KNetwork[A])(
       )
   }
 
-  // Handle findNodes requests...
-  val nodes: ConnectableObservable[(FindNodes[A], Nodes[A] => Task[Unit])] = network.findNodes
-  nodes.foreach {
-    case (findNodesRequest, responseHandler) =>
-      val FindNodes(uuid, nodeRecord, targetNodeId) = findNodesRequest
-
+  network.kRequests.foreach {
+    case (FindNodes(uuid, nodeRecord, targetNodeId), responseHandler) =>
       debug(
         s"Received request FindNodes(${nodeRecord.id.toHex}, $nodeRecord, ${targetNodeId.toHex})"
       )
       add(nodeRecord)
-      val result = embellish(kBuckets.closestNodes(targetNodeId, config.k))
-      responseHandler(Nodes(uuid, config.nodeRecord, result)).runToFuture
-        .onComplete {
-          case Failure(t) =>
-            log.info(
-              s"Nodes response $uuid to ${nodeRecord.routingAddress} failed with exception: $t"
-            )
-          case _ =>
-        }
+
+      val result = Nodes(uuid, config.nodeRecord, embellish(kBuckets.closestNodes(targetNodeId, config.k)))
+
+      sendResponse(result, responseHandler, nodeRecord.routingAddress)
+    case (Ping(uuid, nodeRecord), responseHandler) =>
+      debug(
+        s"Received request Ping(${nodeRecord.id.toHex}, $nodeRecord)"
+      )
+      add(nodeRecord)
+      sendResponse(Pong(uuid, config.nodeRecord), responseHandler, nodeRecord.routingAddress)
   }
-  nodes.connect()
 
   def get(key: BitVector): Future[NodeRecord[A]] = {
     debug(s"get(${key.toHex})")
     getLocally(key).recoverWith { case _ => getRemotely(key) }.recoverWith { case t => giveUp(key, t) }
   }
 
-  def add(nodeRecord: NodeRecord[A]): Unit = {
-    kBuckets.add(nodeRecord.id)
-    nodeRecords.put(nodeRecord.id, nodeRecord)
-    debug(s"Added record (${nodeRecord.id.toHex}, $nodeRecord) to the routing table.")
+  def add(nodeRecord: NodeRecord[A]): Ack = {
+    additions.onNext(nodeRecord)
+  }
+
+  private def sendResponse(response: KResponse[A], responseHandler: Option[KResponse[A]] => Task[Unit], to: A): Unit = {
+    responseHandler(Some(response)).runToFuture
+      .onComplete {
+        case Failure(t) =>
+          log.info(
+            s"Pong response ${response.requestId} to $to failed with exception: $t"
+          )
+        case _ =>
+      }
+  }
+  
+  private def addAll(nodeRecords: Set[NodeRecord[A]]): Future[Ack] = {
+    def loop(records: List[NodeRecord[A]]): Future[Ack] = records match {
+      case Nil =>
+        Continue
+      case h :: t =>
+        add(h).flatMap { _ =>
+          loop(t)
+        }
+    }
+    loop(nodeRecords.toList)
+  }
+
+  private def doAdd: NodeRecord[A] => Future[Ack] = { nodeRecord =>
+    val (iBucket, bucket) = kBuckets.getBucket(nodeRecord.id)
+    debug(s"Handling potential addition of candidate (${nodeRecord.id.toHex}, $nodeRecord) to ibucket $iBucket.")
+    debug(s"iBucket($iBucket) = $bucket")
+    if (bucket.size < config.k) {
+      addNodeRecord(nodeRecord)
+      Continue
```

Comment:
I don't know how why it compiles but it does - but I'm not sure and I'm wondering if that might be some unsound, is there any implicit conversion happening here? @EzequielPostan do you have an idea?

Reply:
monix magic. They defined:

```
monix.execution.Ack

object Continue: Ack.Continue.type extends Ack
```
which looks normal, but then exploring the type of Ack we get
```
monix.execution
abstract sealed class Ack extends Future[Ack]
 with Serializable
```

So, Continue extends Ack and Ack extends Future[Ack]